lib/dry/types/map.rb

```diff
@@ -0,0 +1,110 @@
+module Dry
+  module Types
+    class Map < Definition
+
+      def initialize(keys: Undefined, values: Undefined, meta: EMPTY_HASH)
+        super(::Hash, keys: keys, values: values, meta: meta)
+        validate_options!
+      end
+
+      # @return [Type, Undefined]
+      def keys
+        options[:keys]
+      end
+
+      # @return [Type, Undefined]
+      def values
+        options[:values]
+      end
+
+      # @return [String]
+      def name
+        "Map"
+      end
+
+      # @param [Hash] hash
+      # @return [Hash]
+      def call(hash)
+        try(hash) do |failure|
+          raise MapError, failure.error.join("\n")
+        end.input
+      end
+      alias_method :[], :call
+
+      # @param [Hash] hash
+      # @return [Boolean]
+      def valid?(hash)
+        coerce(hash).success?
+      end
+      alias_method :===, :valid?
+
+      # @param [Hash] hash
+      # @return [Result]
+      def try(hash)
+        result = coerce(hash)
+        return result if result.success? || !block_given?
+        yield(result)
+      end
+
+      # @param meta [Boolean] Whether to dump the meta to the AST
+      # @return [Array] An AST representation
+      def to_ast(meta: true)
+        set = options.reject{ |_,v| Undefined.eql?(v) }
+        set.keys.each do |key|
+          set[key] = set[key].to_ast(meta: meta)
+        end
```

Comment:
This is too abstract for a very specific task: dump `key` and `value` option. Please remove that `Undefined` check and the loop, use `Any` instead of `Undefined`

Reply:
Once set `Any` as default this will be simplified